sympy/combinatorics/rewritingsystem.py

```diff
@@ -80,40 +91,33 @@ def add_rule(self, w1, w2, check=False):
         # redundant rules that would result from processing
         # the overlaps. See [1], Section 3 for details.
 
-        if len(s1) - len(s2) in [0, 1, 2]:
-            if not check and s1 != s2:
-                self.rules[s1] = s2
-            new_keys.add(s1)
+        if len(s1) - len(s2) < 3:
+            if s1 not in self.rules:
+                new_keys.add(s1)
+                if not check:
+                    self._add_rule(s1, s2)
+            if s2**-1 > s1**-1 and s2**-1 not in self.rules:
```

Comment:
In that case it is accounted for by s1 -> s2. The reduction process will substitute for both s1 and s1**-1 except when s2**-1 > s1**-1  so I think that is the only case that needs special treatment.